mgmt/rest/task.go

```diff
@@ -190,62 +193,63 @@ func (s *Server) watchTask(w http.ResponseWriter, r *http.Request, p httprouter.
-			case rbody.TaskWatchMetricEvent, rbody.TaskWatchTaskStarted:
-				// The client can decide to stop receiving on the stream on Task Stopped.
-				// We write the event to the buffer
-				fmt.Fprintf(w, "%s\n", e.ToJSON())
-			case rbody.TaskWatchTaskDisabled, rbody.TaskWatchTaskStopped:
-				// A disabled task should end the streaming and close the connection
-				fmt.Fprintf(w, "%s\n", e.ToJSON())
-				// Flush since we are sending nothing new
+		case <-n:
+			respond(200, &rbody.ScheduledTaskWatchingEnded{}, w)
+			return
+		default:
+			msg, err := stream.Recv()
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				se := rbody.StreamedTaskEvent{
+					EventType: rpc.Watch_ERROR,
+					Message:   grpc.ErrorDesc(err),
+				}
+				fmt.Fprintf(w, "%s\n", se.ToJSON())
 				flusher.Flush()
-				// Close out watcher removing it from the scheduler
-				tc.Close()
-				// exit since this client is no longer listening
-				respond(200, &rbody.ScheduledTaskWatchingEnded{}, w)
+				break
+			}
+			if time.Now().Sub(t).Seconds() < StreamingBufferWindow {
+				break
 			}
-			// If we are at least above our minimum buffer time we flush to send
-			if time.Now().Sub(t).Seconds() > StreamingBufferWindow {
+			t = time.Now()
+			switch msg.EventType {
+			case rpc.Watch_METRICS_COLLECTED, rpc.Watch_TASK_STARTED:
+				logger.WithField("msg_type", msg.EventType.String()).Debug(msg.ToJson())
+				fmt.Fprintf(w, "%s\n", msg.ToJson())
+				flusher.Flush()
+			case rpc.Watch_TASK_DISABLED, rpc.Watch_TASK_STOPPED:
+				logger.WithField("msg_type", msg.EventType.String()).Debug(msg.ToJson())
+				fmt.Fprintf(w, "%s\n", msg.ToJson())
 				flusher.Flush()
-				t = time.Now()
 			}
-		case <-n:
-			logger.WithFields(log.Fields{
-				"task-id": id,
-			}).Debug("client disconnecting")
-			// Flush since we are sending nothing new
-			flusher.Flush()
-			// Close out watcher removing it from the scheduler
-			tc.Close()
-			// exit since this client is no longer listening
-			respond(200, &rbody.ScheduledTaskWatchingEnded{}, w)
 		}
 	}
 }
 
 func (s *Server) startTask(w http.ResponseWriter, r *http.Request, p httprouter.Params) {
 	id := p.ByName("id")
-	errs := s.mt.StartTask(id)
-	if errs != nil {
-		if strings.Contains(errs[0].Error(), ErrTaskNotFound.Error()) {
-			respond(404, rbody.FromSnapErrors(errs), w)
+	reply, err := s.mt.StartTask(context.Background(), &rpc.StartTaskArg{Id: id})
+	if err != nil {
+		respond(404, rbody.FromError(err), w)
+		return
+	}
+
+	if reply.Errors != nil {
+		if strings.Contains(reply.Errors[0].ErrorString, ErrTaskNotFound.Error()) {
+			respond(404, rbody.FromSnapErrors(rpc.ConvertSnapErrors(reply.Errors)), w)
 			return
 		}
-		if strings.Contains(errs[0].Error(), ErrTaskDisabledNotRunnable.Error()) {
-			respond(409, rbody.FromSnapErrors(errs), w)
+		if strings.Contains(reply.Errors[0].ErrorString, ErrTaskDisabledNotRunnable.Error()) {
+			respond(404, rbody.FromSnapErrors(rpc.ConvertSnapErrors(reply.Errors)), w)
```

Comment:
Another option is `409`:

> _409 Conflict_
>
> The request could not be completed due to a conflict with the current state of the resource. This code is only allowed in situations where it is expected that the user might be able to resolve the conflict and resubmit the request. The response body SHOULD include enough information for the user to recognize the source of the conflict. Ideally, the response entity would include enough information for the user or user agent to fix the problem; however, that might not be possible and is not required.

Source: https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html